cub/test/catch2_test_device_transform.cu

```diff
@@ -0,0 +1,483 @@
+  c2h::device_vector<type> a(num_items);
+  c2h::device_vector<type> b(num_items);
+  thrust::sequence(a.begin(), a.end());
+  thrust::sequence(b.begin(), b.end(), 42);
+
+  c2h::device_vector<type> result(num_items);
+  transform_many_stable(
+    ::cuda::std::make_tuple(thrust::raw_pointer_cast(a.data())),
+    result.begin(),
+    num_items,
+    plus_needs_stable_address{thrust::raw_pointer_cast(a.data()), thrust::raw_pointer_cast(b.data())});
+
+  // compute reference and verify
+  c2h::device_vector<type> a_h = a;
+  c2h::device_vector<type> b_h = b;
+  c2h::host_vector<type> reference_h(num_items);
+  std::transform(a_h.begin(), a_h.end(), b_h.begin(), reference_h.begin(), std::plus<type>{});
+  REQUIRE(reference_h == result);
+}
+
+// Non-trivially-copyable/relocatable type which cannot be copied using std::memcpy or cudaMemcpy
+struct non_trivial
+{
+  int data;
+
+  non_trivial() = default;
+
+  _CCCL_HOST_DEVICE explicit non_trivial(int data)
+      : data(data)
+  {}
+
+  _CCCL_HOST_DEVICE non_trivial(const non_trivial& nt)
+      : data(nt.data)
+  {}
+
+  _CCCL_HOST_DEVICE auto operator=(const non_trivial& nt) -> non_trivial&
+  {
+    data = nt.data;
+    return *this;
+  }
+
+  _CCCL_HOST_DEVICE auto operator-() const -> non_trivial
+  {
+    return non_trivial{-data};
+  }
+
+  friend _CCCL_HOST_DEVICE auto operator==(non_trivial a, non_trivial b) -> bool
+  {
+    return a.data == b.data;
+  }
+};
+static_assert(!::cuda::std::is_trivially_copyable<non_trivial>::value, ""); // as required by the standard
+static_assert(!thrust::is_trivially_relocatable<non_trivial>::value, ""); // CUB uses this check internally
+
+// Note(bgruber): I gave up on writing a test that checks whether the copy ctor/assignment operator is actually called
+// (e.g. by tracking/counting invocations of those), since C++ allows (but not guarantees) elision of these operations.
+// Also thrust algorithms perform a lot of copies in-between, so the test needs to use only raw allocations and
+// iteration for setup and checking.
+CUB_TEST("DeviceTransform::Transform not trivially relocatable", "[device][device_transform]")
+{
+  constexpr int num_items = 100;
+  c2h::device_vector<non_trivial> input(num_items, non_trivial{42});
+  c2h::device_vector<non_trivial> result(num_items);
+  transform_many(::cuda::std::make_tuple(input.begin()), result.begin(), num_items, ::cuda::std::negate<>{});
+
+  const auto reference = c2h::device_vector<non_trivial>(num_items, non_trivial{-42});
+  REQUIRE((reference == result));
+}
+
+CUB_TEST("DeviceTransform::Transform buffer start alignment",
+         "[device][device_transform]",
+         c2h::type_list<std::uint8_t, std::uint16_t, float, double>)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  constexpr int num_items = 1000;
+  const int offset        = GENERATE(1, 2, 4, 8, 16, 32, 64, 128); // global memory is always at least 256 byte aligned
+  c2h::device_vector<type> input(num_items, 42);
```

Comment:
important: I don't think that this input leads to a sound test. If we accidentally transform prelast element and store the result in result + offset, we woudln't notice. Let's have a `thrust::sequence` instead of constant input.

Reply:
Good reasoning!